File: app-mvp/src/main/java/com/xuie/imaginaryandroid/gui/MainActivity.java
package xuk.imaginary.gui;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.view.MenuItem;

import xuk.imaginary.R;
import xuk.imaginary.gui.gank.meizhi.MeizhiFragment;
import xuk.imaginary.gui.nets.news.NewsListFragment;
import xuk.imaginary.gui.nets.video.VideoMainFragment;
import xuk.imaginary.gui.settings.SettingsFragment;

import butterknife.BindView;
import butterknife.ButterKnife;
import fm.jiecao.jcvideoplayer_lib.JCVideoPlayer;

/**
 * @author xuie
 */
public class MainActivity extends BaseActivity {
    private static final String TAG = "MainActivity";
    @BindView(R.id.view_pager) ViewPager viewPager;
    @BindView(R.id.navigation) BottomNavigationView navigation;

    private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = new BottomNavigationView.OnNavigationItemSelectedListener() {

        @Override
        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
            switch (item.getItemId()) {
                case R.id.navigation_dashboard:
                    viewPager.setCurrentItem(0);
                    return true;
                case R.id.navigation_video:
                    viewPager.setCurrentItem(1);
                    return true;
                case R.id.navigation_gank:
                    viewPager.setCurrentItem(2);
                    return true;
                case R.id.navigation_settings:
                default:
                    viewPager.setCurrentItem(3);
                    return true;
            }
        }

    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        Log.d(TAG, "onCreate");

        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);
        viewPager.setAdapter(new MyViewPagerAdapter(getSupportFragmentManager()));
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            boolean isUser = false;

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                if (isUser) {
                    int itemId = navigation.getMenu().getItem(position).getItemId();
                    navigation.setSelectedItemId(itemId);
                    isUser = false;
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                if (state == ViewPager.SCROLL_STATE_DRAGGING) {
                    isUser = true;
                }
            }
        });
        navigation.setSelectedItemId(navigation.getMenu().getItem(1).getItemId());
    }

    private class MyViewPagerAdapter extends FragmentPagerAdapter {

        MyViewPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    return NewsListFragment.getInstance();
                case 1:
                    return VideoMainFragment.getInstance();
                case 2:
                    return MeizhiFragment.getInstance();
                case 3:
                default:
                    return SettingsFragment.getInstance();
            }
        }

        @Override
        public int getCount() {
            return 4;
        }
    }

    @Override
    public void onBackPressed() {
        if (JCVideoPlayer.backPress()) {
            return;
        }
        super.onBackPressed();
    }
}

File: app-mvvm-kotlin/src/main/kotlin/xuk/imaginary/gui/MainActivity.kt
package xuk.imaginary.gui

import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.lifecycle.ViewModelProviders
import androidx.viewpager.widget.ViewPager
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.xuie.imaginary.R
import com.xuie.imaginary.ViewModelFactory
import com.xuie.imaginary.base.BaseActivity
import com.xuie.imaginary.databinding.ActivityMainBinding
import com.xuie.imaginary.gui.gank.meizhi.MeiZhiViewModule
import com.xuie.imaginary.gui.gank.meizhi.MeizhiFragment
import com.xuie.imaginary.gui.nets.news.NewsListFragment
import com.xuie.imaginary.gui.nets.news.NewsListViewModule
import com.xuie.imaginary.gui.nets.video.VideoMainFragment
import com.xuie.imaginary.gui.nets.video.VideosViewModule
import fm.jiecao.jcvideoplayer_lib.JCVideoPlayer

/**
 * @author Jie Xu
 */
class MainActivity : BaseActivity() {
  private val mOnNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item ->
    when (item.itemId) {
      R.id.navigation_dashboard -> {
        mBinding.viewPager.currentItem = 0
        true
      }
      R.id.navigation_video -> {
        mBinding.viewPager.currentItem = 1
        true
      }
      R.id.navigation_gank -> {
        mBinding.viewPager.currentItem = 2
        true
      }
//      R.id.navigation_settings -> {
//        mBinding.viewPager.currentItem = 3
//        true
//      }
      else -> true
    }
  }

  private lateinit var mBinding: ActivityMainBinding

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
    mBinding.navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)
    mBinding.viewPager.adapter = MyViewPagerAdapter(supportFragmentManager)
    mBinding.viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
      var isUser = false

      override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

      override fun onPageSelected(position: Int) {
        if (isUser) {
          val itemId = mBinding.navigation.menu.getItem(position).itemId
          mBinding.navigation.selectedItemId = itemId
          isUser = false
        }
      }

      override fun onPageScrollStateChanged(state: Int) {
        if (state == ViewPager.SCROLL_STATE_DRAGGING) {
          isUser = true
        }
      }
    })
    mBinding.navigation.selectedItemId = mBinding.navigation.menu.getItem(1).itemId
  }

  private inner class MyViewPagerAdapter internal constructor(fm: FragmentManager) : FragmentPagerAdapter(fm) {

    override fun getItem(position: Int): Fragment {
      return when (position) {
        0 -> NewsListFragment.instance
        1 -> VideoMainFragment.instance
        2 -> MeizhiFragment.instance
//        3 -> SettingsFragment.instance
        else -> MeizhiFragment.instance
      }
    }

    override fun getCount(): Int {
      return 3
    }
  }

  override fun onBackPressed() {
    if (JCVideoPlayer.backPress()) {
      return
    }
    super.onBackPressed()
  }

  companion object {

    fun obtainViewModel(activity: FragmentActivity): MeiZhiViewModule {
      // Use a Factory to inject dependencies into the ViewModel
      val factory = ViewModelFactory.getInstance(activity.application)
      return ViewModelProviders.of(activity, factory).get(MeiZhiViewModule::class.java)
    }

    fun obtainNewsViewModel(activity: FragmentActivity): NewsListViewModule {
      val factory = ViewModelFactory.getInstance(activity.application)
      return ViewModelProviders.of(activity, factory).get(NewsListViewModule::class.java)
    }

    fun obtainVideosViewModel(activity: FragmentActivity): VideosViewModule {
      val factory = ViewModelFactory.getInstance(activity.application)
      return ViewModelProviders.of(activity, factory).get(VideosViewModule::class.java)
    }
  }
}
